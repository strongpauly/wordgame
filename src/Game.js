import React, { Component } from 'react';
import Cell from './Cell';
import PropTypes from 'prop-types';
import WordSet from './model/WordSet';

export default class Game extends Component {

  static propTypes = {
    width: PropTypes.number,
    height: PropTypes.number,
    onRestart: PropTypes.func
  }

  selecting = false;

  constructor(props) {
    super(props);
    this.words = new WordSet();
    this.state = {
      completed: false,
      selected: new Set(),
      charList: []
    };
  }

  getCellKey(x, y) {
    return x + ',' + y;
  }

  getCellCoord(key) {
    let split = key.split(',');
    return {x:parseInt(split[0], 10), y: parseInt(split[1], 10)};
  }

  startTimer() {
    if(!this.timerId) {
      this.setState({time: 1});
      this.timerId = setInterval(() => {
        this.setState((previousState) => ({
          time:previousState.time + 1
        }));
      }, 1000);
    }
  }

  stopTimer() {
    if(this.timerId) {
      clearInterval(this.timerId);
    }
  }

  componentWillUnmount() {
    this.stopTimer();
    window.removeEventListener('mouseup', this.onSelectEnd);
  }

  onCheck(cellX, cellY) {
    if(!this.state.completed) {
      this.startTimer();
      let key = this.getCellKey(cellX, cellY);
      if(this.state.mines.has(key)) {
        this.setState({exploded:true, completed:true});
        this.stopTimer();
      } else {
        this.doCheck(key);
        if(this.hasWon()) {
          this.setState({completed:true});
          this.stopTimer();
        }
        this.setState({
          checked: this.state.checked,
          marked: this.state.marked
        });
      }
    }
  }

  restart = () => {
    this.props.onRestart();
  }

  hasWon() {
    return false;
  }

  onSelectStart = (x, y, char) => {
    this.selecting = true;
    window.addEventListener('mouseup', this.onSelectEnd, false);
    let selected = new Set();
    selected.add(this.getCellKey(x, y));
    this.setState({
      selected: selected,
      charList: [char]
    });
  }

  onSelectOver = (x, y, char) => {
    if(this.selecting) {
      let newSelected = this.state.selected;
      newSelected.add(this.getCellKey(x, y));
      this.setState({
        selected: newSelected,
        charList: this.state.charList.concat(char)
      });
    }
  }

  onSelectEnd = () => {
    if(this.selecting) {
      this.selecting = false;
      window.removeEventListener('mouseup', this.onSelectEnd);
    }
  }

  render() {
    let widthArray = new Array(this.props.width).fill();
    let heightArray = new Array(this.props.height).fill();
    let cells = heightArray.map((emptyY, y) => {
      let row = widthArray.map((emptyX, x) => {
        let key = this.getCellKey(x, y);
        return <Cell key={key} x={x} y={y} char={this.words.getCharAt(x, y)}
          selected={this.state.selected.has(key)}
          onSelectStart={this.onSelectStart}
          onSelectOver={this.onSelectOver}
          onSelectEnd={this.onSelectEnd}></Cell>;
      });
      return <tr className="row" key={'row' + y}>{row}</tr>;
    });
    return <div className="gridContainer">
            <div>
                <div className="header">
                    <div className="timer">{this.state.time || ' '}</div>
                </div>
                <table className={this.state.completed ? 'grid completed' : 'grid'}>
                    <tbody>{cells}</tbody>
                </table>
            </div>
            <div>{this.state.charList.join('')}</div>
        </div>;
  }
}
