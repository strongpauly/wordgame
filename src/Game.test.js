import React from 'react';
import Game from './Game';
import {shallow} from 'enzyme';

/* eslint-env jest */

describe('<Game>', () => {
  jest.useFakeTimers();
  it('renders without crashing', () => {
    shallow(<Game width={10} height={10}/>);
  });

  it('has header', () => {
    const game = shallow(<Game width={0} height={0}/>);
    expect(game).toMatchSnapshot();
    expect(game.find('.header')).toHaveLength(1);
  });

  it('creates grid based on width and height', () => {
    const game = shallow(<Game width={2} height={2}/>);
    expect(game).toMatchSnapshot();
    expect(game.find('tr')).toHaveLength(2); //Height 2.
    expect(game.find('Cell')).toHaveLength(4); //Height 2 x Width 2.
  });

  // it('has completed class when game is won', () => {
  //   const game = shallow(<Game width={1} height={1}/>);
  //   expect(game).toMatchSnapshot();
  //   const completed = game.find('.completed');
  //   expect(completed).toHaveLength(1);
  // });

  // it('will call restart handler when status button is clicked', () => {
  //   const onRestart = sinon.spy();
  //   const game = mount(<Game width={30} height={30} numMines={1} onRestart={onRestart}/>);
  //   game.find('.status').simulate('click');
  //   expect(onRestart.calledOnce).toEqual(true);
  // });

  // it('will update timer after game starts', () => {
  //   const game = mount(<Game width={3} height={3} numMines={2}/>);
  //   const timer = game.find('.header .timer');
  //   expect(timer).toHaveLength(1);
  //   expect(timer.text()).toEqual(' ');
  //   game.find('Cell td').at(0).simulate('contextMenu');
  //   expect(timer.text()).toEqual('1');
  //   jest.runTimersToTime(1000);
  //   expect(timer.text()).toEqual('2');
  // });
  //
  // it('wont update timer after game completed', () => {
  //   const game = mount(<Game width={1} height={1} numMines={1}/>);
  //   const timer = game.find('.header .timer');
  //   expect(timer.text()).toEqual(' ');
  //   game.find('Cell td').at(0).simulate('contextMenu');
  //   expect(timer.text()).toEqual(' ');
  //   jest.runTimersToTime(1000);
  //   expect(timer.text()).toEqual(' ');
  // });

});
