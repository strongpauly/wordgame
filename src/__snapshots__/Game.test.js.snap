// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game> creates grid based on width and height 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="gridContainer"
>
    <div>
        <div
            className="header"
        >
            <div
                className="timer"
            >
                 
            </div>
        </div>
        <table
            className="grid"
        >
            <tbody>
                <tr
                    className="row"
                >
                    <Cell
                        char="K"
                        onSelectEnd={[Function]}
                        onSelectOver={[Function]}
                        onSelectStart={[Function]}
                        selected={false}
                        x={0}
                        y={0}
                    />
                    <Cell
                        char="R"
                        onSelectEnd={[Function]}
                        onSelectOver={[Function]}
                        onSelectStart={[Function]}
                        selected={false}
                        x={1}
                        y={0}
                    />
                </tr>
                <tr
                    className="row"
                >
                    <Cell
                        char="I"
                        onSelectEnd={[Function]}
                        onSelectOver={[Function]}
                        onSelectStart={[Function]}
                        selected={false}
                        x={0}
                        y={1}
                    />
                    <Cell
                        char="E"
                        onSelectEnd={[Function]}
                        onSelectOver={[Function]}
                        onSelectStart={[Function]}
                        selected={false}
                        x={1}
                        y={1}
                    />
                </tr>
            </tbody>
        </table>
    </div>
    <div />
</div>,
  "nodes": Array [
    <div
      className="gridContainer"
>
      <div>
            <div
                  className="header"
            >
                  <div
                        className="timer"
                  >
                         
                  </div>
            </div>
            <table
                  className="grid"
            >
                  <tbody>
                        <tr
                              className="row"
                        >
                              <Cell
                                    char="K"
                                    onSelectEnd={[Function]}
                                    onSelectOver={[Function]}
                                    onSelectStart={[Function]}
                                    selected={false}
                                    x={0}
                                    y={0}
                              />
                              <Cell
                                    char="R"
                                    onSelectEnd={[Function]}
                                    onSelectOver={[Function]}
                                    onSelectStart={[Function]}
                                    selected={false}
                                    x={1}
                                    y={0}
                              />
                        </tr>
                        <tr
                              className="row"
                        >
                              <Cell
                                    char="I"
                                    onSelectEnd={[Function]}
                                    onSelectOver={[Function]}
                                    onSelectStart={[Function]}
                                    selected={false}
                                    x={0}
                                    y={1}
                              />
                              <Cell
                                    char="E"
                                    onSelectEnd={[Function]}
                                    onSelectOver={[Function]}
                                    onSelectStart={[Function]}
                                    selected={false}
                                    x={1}
                                    y={1}
                              />
                        </tr>
                  </tbody>
            </table>
      </div>
      <div />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance":  {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Game
        height={2}
        width={2}
/>,
      "_debugID": 5,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Game {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "onSelectEnd": [Function],
        "onSelectOver": [Function],
        "onSelectStart": [Function],
        "props": Object {
          "height": 2,
          "width": 2,
        },
        "refs": Object {},
        "restart": [Function],
        "selecting": false,
        "state": Object {
          "charList": Array [],
          "completed": false,
          "selected": Set {},
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
        "words": WordSet {
          "characters": Array [
            Array [
              "K",
              "I",
              "L",
              "L",
              "E",
              "R",
            ],
            Array [
              "R",
              "E",
              "S",
              "A",
              "L",
              "E",
            ],
            Array [
              "H",
              "O",
              "R",
              "R",
              "O",
              "R",
            ],
            Array [
              "M",
              "I",
              "R",
              "R",
              "O",
              "R",
            ],
            Array [
              "G",
              "O",
              "O",
              "G",
              "L",
              "E",
            ],
            Array [
              "H",
              "A",
              "N",
              "D",
              "E",
              "D",
            ],
          ],
        },
      },
      "_mountOrder": 3,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="gridContainer"
>
          <div>
                    <div
                              className="header"
                    >
                              <div
                                        className="timer"
                              >
                                         
                              </div>
                    </div>
                    <table
                              className="grid"
                    >
                              <tbody>
                                        <tr
                                                  className="row"
                                        >
                                                  <Cell
                                                            char="K"
                                                            onSelectEnd={[Function]}
                                                            onSelectOver={[Function]}
                                                            onSelectStart={[Function]}
                                                            selected={false}
                                                            x={0}
                                                            y={0}
                                                  />
                                                  <Cell
                                                            char="R"
                                                            onSelectEnd={[Function]}
                                                            onSelectOver={[Function]}
                                                            onSelectStart={[Function]}
                                                            selected={false}
                                                            x={1}
                                                            y={0}
                                                  />
                                        </tr>
                                        <tr
                                                  className="row"
                                        >
                                                  <Cell
                                                            char="I"
                                                            onSelectEnd={[Function]}
                                                            onSelectOver={[Function]}
                                                            onSelectStart={[Function]}
                                                            selected={false}
                                                            x={0}
                                                            y={1}
                                                  />
                                                  <Cell
                                                            char="E"
                                                            onSelectEnd={[Function]}
                                                            onSelectOver={[Function]}
                                                            onSelectStart={[Function]}
                                                            selected={false}
                                                            x={1}
                                                            y={1}
                                                  />
                                        </tr>
                              </tbody>
                    </table>
          </div>
          <div />
</div>,
        "_debugID": 6,
        "_renderedOutput": <div
          className="gridContainer"
>
          <div>
                    <div
                              className="header"
                    >
                              <div
                                        className="timer"
                              >
                                         
                              </div>
                    </div>
                    <table
                              className="grid"
                    >
                              <tbody>
                                        <tr
                                                  className="row"
                                        >
                                                  <Cell
                                                            char="K"
                                                            onSelectEnd={[Function]}
                                                            onSelectOver={[Function]}
                                                            onSelectStart={[Function]}
                                                            selected={false}
                                                            x={0}
                                                            y={0}
                                                  />
                                                  <Cell
                                                            char="R"
                                                            onSelectEnd={[Function]}
                                                            onSelectOver={[Function]}
                                                            onSelectStart={[Function]}
                                                            selected={false}
                                                            x={1}
                                                            y={0}
                                                  />
                                        </tr>
                                        <tr
                                                  className="row"
                                        >
                                                  <Cell
                                                            char="I"
                                                            onSelectEnd={[Function]}
                                                            onSelectOver={[Function]}
                                                            onSelectStart={[Function]}
                                                            selected={false}
                                                            x={0}
                                                            y={1}
                                                  />
                                                  <Cell
                                                            char="E"
                                                            onSelectEnd={[Function]}
                                                            onSelectOver={[Function]}
                                                            onSelectStart={[Function]}
                                                            selected={false}
                                                            x={1}
                                                            y={1}
                                                  />
                                        </tr>
                              </tbody>
                    </table>
          </div>
          <div />
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Game
    height={2}
    width={2}
/>,
}
`;

exports[`<Game> has header 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="gridContainer"
>
    <div>
        <div
            className="header"
        >
            <div
                className="timer"
            >
                 
            </div>
        </div>
        <table
            className="grid"
        >
            <tbody>
                
            </tbody>
        </table>
    </div>
    <div />
</div>,
  "nodes": Array [
    <div
      className="gridContainer"
>
      <div>
            <div
                  className="header"
            >
                  <div
                        className="timer"
                  >
                         
                  </div>
            </div>
            <table
                  className="grid"
            >
                  <tbody>
                        
                  </tbody>
            </table>
      </div>
      <div />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance":  {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Game
        height={0}
        width={0}
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Game {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "onSelectEnd": [Function],
        "onSelectOver": [Function],
        "onSelectStart": [Function],
        "props": Object {
          "height": 0,
          "width": 0,
        },
        "refs": Object {},
        "restart": [Function],
        "selecting": false,
        "state": Object {
          "charList": Array [],
          "completed": false,
          "selected": Set {},
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
        "words": WordSet {
          "characters": Array [
            Array [
              "K",
              "I",
              "L",
              "L",
              "E",
              "R",
            ],
            Array [
              "R",
              "E",
              "S",
              "A",
              "L",
              "E",
            ],
            Array [
              "H",
              "O",
              "R",
              "R",
              "O",
              "R",
            ],
            Array [
              "M",
              "I",
              "R",
              "R",
              "O",
              "R",
            ],
            Array [
              "G",
              "O",
              "O",
              "G",
              "L",
              "E",
            ],
            Array [
              "H",
              "A",
              "N",
              "D",
              "E",
              "D",
            ],
          ],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="gridContainer"
>
          <div>
                    <div
                              className="header"
                    >
                              <div
                                        className="timer"
                              >
                                         
                              </div>
                    </div>
                    <table
                              className="grid"
                    >
                              <tbody>
                                        
                              </tbody>
                    </table>
          </div>
          <div />
</div>,
        "_debugID": 4,
        "_renderedOutput": <div
          className="gridContainer"
>
          <div>
                    <div
                              className="header"
                    >
                              <div
                                        className="timer"
                              >
                                         
                              </div>
                    </div>
                    <table
                              className="grid"
                    >
                              <tbody>
                                        
                              </tbody>
                    </table>
          </div>
          <div />
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Game
    height={0}
    width={0}
/>,
}
`;
